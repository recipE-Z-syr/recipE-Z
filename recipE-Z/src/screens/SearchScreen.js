import React, { Component } from 'react';
import { Text, TextInput, View, TouchableOpacity, Dimensions, Image } from 'react-native';
import defaultStyles from './stylesheet';
import Autocomplete from 'react-native-autocomplete-input';
import NavigationBar from './NavigationBar';

class IngredientItem extends React.Component {
  constructor(props) {
    super(props);
    this.state = {isClicked: false};
  }

  render () {
    const {item, removeHandler} = this.props;
    return (
      <View>
        <TouchableOpacity style={defaultStyles.ingredient} onPress={() => {this.setState({isClicked: !this.state.isClicked})}}>
          <Text key={item} style={defaultStyles.ingredientText}>{item}</Text>
        </TouchableOpacity>
        <TouchableOpacity onPress={() => removeHandler(item)}>
          <Image
          key={item}
          style={[{width: 20, height: 20, zIndex: 2, marginTop: -60, marginLeft: -5}, {display: this.state.isClicked ? 'flex' : 'none'}]}
          source={require('../img/delete.png')}
          />
        </TouchableOpacity>
      </View>
    )
  }
}

class SearchScreen extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      ingredients: [],
      ingdata: ['eggs', 'milk', 'cheese', 'carrots', 'ice cream', 'cheddar', 'flour', 'macaroni', 'pasta', 'broth', 'chicken', 'beef', 'lettuce', 'peas'],
      query: '',
      cuisine: '',
      allergies: [],
      exclusions: []
    };

    this.handleIngredients = this.handleIngredients.bind(this);
    this.handleQuery = this.handleQuery.bind(this); //query = dish (eg pasta)
    this.handleCuisine = this.handleCuisine.bind(this);
    this.handleAllergies = this.handleAllergies.bind(this);
    this.handleExclusions = this.handleExclusions.bind(this);
    //this.handleMaxCals = this.handleMaxCals.bind(this); //maxCals does not work when no value is input - it breaks the whole thing
    this.handleExclusions = this.handleExclusions.bind(this);
    this.suggestIngredient = this.suggestIngredient.bind(this);
    this.clear = this.clear.bind(this);

  }

  // Handler methods for text inputs
  handleQuery(text) {
    this.setState({ query: text });
  }

  handleCuisine(text) {
    this.setState({ cuisine: text });
  }

  handleAllergies(text) {
    this.setState({ allergies: [...this.state.allergies, text] });
  }

  handleExclusions(text) {
    this.setState({ exclusions: [...this.state.exclusions, text] });
  }

  handleIngredients(text) {
    if(!this.state.ingredients.includes(text)){
      this.setState({ ingredients: [...this.state.ingredients, text] });
      this.setState({ query: '' });
    }
  }

  removeItem(item) {
    // console.log(this.state.ingredients);
    var array = this.state.ingredients;
    var index = array.indexOf(item);
    if (index !== -1) array.splice(index, 1);
    this.setState({ingredients: array})
  }

  clear() {
    this.setState({ingredients: []});
  }

  suggestIngredient(text) {
    if (text === '') {
      return [];
    }
    const { ingdata } = this.state;
    const regex = new RegExp(`${text.trim()}`, 'i');
    return ingdata.filter(ing => ing.search(regex) >= 0);
  }

  //////////// render block //////////////////

  render() {
    const {navigation} = this.props;
    const { query } = this.state;
    const data = this.suggestIngredient(query);
    return (
      <View style={{
        flex:1,
        height: Dimensions.get('window').height,
        backgroundColor:'#fff',
        alignItems: 'center',
        justifyContent: 'space-around'
      }}>

       <View style= {[defaultStyles.searchContainer]}>
       <NavigationBar navigation={this.props.navigation}/>
       <Image
        style={{width: 65, height: 65, marginBottom: 40}}
        source={require('../img/logo.png')}
        />
       <Text style={[defaultStyles.h1, {fontSize: 24}]}>Search for recipes</Text>
       <Text style={[defaultStyles.textLink, {textAlign: 'left', paddingTop: 5}]}>You can search by entering the ingredients you have</Text>
       <Text style={[defaultStyles.textLink, {textAlign: 'left', paddingBottom: 15}]}>or by typing in a recipe name</Text>
        <Image
        style={{height: 20, width:20, position: 'absolute', top: 217, right: 50, zIndex: 2}}
        source={require('../img/search.png')}
        />
        <Autocomplete
          containerStyle={defaultStyles.autocompleteContainer}
          inputContainerStyle={defaultStyles.acInputContainer}
          listStyle={{borderWidth: 0}}
          data={data}
          defaultValue={query}
          onChangeText={text => this.setState({ query: text })}
          renderItem={({ item, i }) => (
            <TouchableOpacity key={item} style={defaultStyles.acListStyle} onPress={() => {this.handleIngredients(item)}}>
              <Image
              style={{height: 18, width:18, marginRight: 8}}
              source={require('../img/plus.png')}
              />
              <Text style={{color: '#ed4848'}}>{item}</Text>
            </TouchableOpacity>
          )}
        />
        </View>
        <View style = {[defaultStyles.inputContainer,]}>
          <Text style={[defaultStyles.h1, {fontSize: 24, color: "#fff", width: 350}]}>Ingredients:</Text>
          <TouchableOpacity onPress = {this.clear} style={{position: 'absolute', top: 30, right: 40}}>
            <Text style = {[defaultStyles.textLink, {fontSize: 14, color: '#fff'}]}>CLEAR</Text>
          </TouchableOpacity>
          <View style= {defaultStyles.ingredientsContainer}>
          {this.state.ingredients.map(item => (
            <IngredientItem item={item} removeHandler={this.removeItem.bind(this)}/>
          ))}
          </View>
          <TouchableOpacity
            style = {[defaultStyles.redButton, {backgroundColor: '#fff', width: 350}]}
            onPress ={() => this.props.navigation.navigate(
              'Results',
              { ingArray: this.state.ingredients })}
            >
            <Text style = {[defaultStyles.redButtonText, {color: '#ed4848'}]}>Find Recipes</Text>
          </TouchableOpacity>
        </View>
     </View>
   ); //end return
 }//end render
}//end class

export default SearchScreen;
