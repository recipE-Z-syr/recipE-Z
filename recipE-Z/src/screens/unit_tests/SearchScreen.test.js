//SearchScreen.test.js

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, ShallowWrapper } from "enzyme";

import SearchScreen from '../SearchScreen';

describe('<SearchScreen />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<SearchScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should change state if query is entered', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleQuery('someQuery');
  	expect(instanceOf.state.query).toEqual('someQuery');
  });
  it('should change state if cuisine is entered', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleCuisine('someCuisine');
  	expect(instanceOf.state.cuisine).toEqual('someCuisine');
  });
  it('should change state if allergies is entered', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleAllergies('someAllergy1');
  	expect(instanceOf.state.allergies).toContain('someAllergy1');
  	instanceOf.handleAllergies('someAllergy2');
  	expect(instanceOf.state.allergies).toContain('someAllergy2');
  });
  it('should change state if exclusions is entered', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleExclusions('someExclusion1');
  	expect(instanceOf.state.exclusions).toContain('someExclusion1');
  	instanceOf.handleExclusions('someExclusion2');
  	expect(instanceOf.state.exclusions).toContain('someExclusion2');
  });
  it('should change state if ingredients is entered', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleIngredients('someIng1');
  	expect(instanceOf.state.ingredients).toContain('someIng1');
  	expect(instanceOf.state.query).toEqual('');
  	instanceOf.handleIngredients('someIng2');
  	expect(instanceOf.state.ingredients).toContain('someIng2');
  	expect(instanceOf.state.query).toEqual('');
  });
  it('should not change state if duplicate ingredient is entered', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleIngredients('someIng1');
  	expect(instanceOf.state.ingredients).toContain('someIng1');
  	expect(instanceOf.state.query).toEqual('');
  	instanceOf.handleIngredients('someIng1');
  	expect(instanceOf.state.ingredients).toContain('someIng1');
  });
  it('should not change state if duplicate ingredient is entered', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleIngredients('abc');
  	expect(instanceOf.state.ingredients).toContain('abc');
  	expect(instanceOf.state.query).toEqual('');
  });
  it('should remove ingredient', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleIngredients('item1');
  	instanceOf.removeItem('item1');
  	expect(instanceOf.state.ingredients).toEqual([]);
  });
  it('should not remove ingredient if not in state', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleIngredients('item1');
  	instanceOf.handleIngredients('item2');
  	instanceOf.removeItem('item3');
  	expect(instanceOf.state.ingredients).toEqual(['item1', 'item2']);
  });
  it('should remove only one ingredient', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.handleIngredients('item1');
  	instanceOf.handleIngredients('item2');
  	instanceOf.handleIngredients('item3');
  	instanceOf.removeItem('item2');
  	expect(instanceOf.state.ingredients).toEqual(['item1', 'item3']);
  });
  it('should clear ingredients', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	instanceOf.clear();
  	expect(instanceOf.state.ingredients).toEqual([]);
  });
  it('should suggest ingredients containing letter', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	expect(instanceOf.suggestIngredient('m')).toEqual(['milk', 'ice cream', 'macaroni']);
  });
  it('should suggest ingredients containing phrase', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	expect(instanceOf.suggestIngredient('che')).toEqual(['cheese', 'cheddar']);
  });
  it('should suggest no ingredients if empty', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	expect(instanceOf.suggestIngredient('')).toEqual([]);
  });
  it('should suggest no ingredients if not in list', () => {
  	const instanceOf = renderer.create(<SearchScreen />).getInstance();
  	expect(instanceOf.suggestIngredient('notInList')).toEqual([]);
  });
});

const createTestProps = () => ({
  navigation: {
    navigate: jest.fn()
  }
});

// test('Navigates to Home', () => {
//   // const spy = jest.spyOn(navigation, 'navigate')
//   let wrapper: ShallowWrapper;
//   let props: any
//   props = createTestProps();
//   wrapper = shallow(<SearchScreen {...props} />);  

//   const startButton = wrapper.find('TouchableOpacity').at(0)
//   startButton.props().onPress()

//   expect(props.navigation.navigate).toBeCalledWith('Search')
// })
