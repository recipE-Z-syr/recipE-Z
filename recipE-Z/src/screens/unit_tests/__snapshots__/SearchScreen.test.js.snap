// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "height": 1334,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#fff",
          "flex": 5,
          "justifyContent": "center",
          "paddingLeft": 32,
          "paddingRight": 32,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row-reverse",
          "position": "absolute",
          "right": 30,
          "top": 30,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 25,
              "marginLeft": 20,
              "width": 25,
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 25,
              "marginLeft": 20,
              "width": 25,
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 25,
              "marginLeft": 20,
              "width": 25,
            }
          }
        />
      </View>
    </View>
    <Image
      source={1}
      style={
        Object {
          "height": 65,
          "marginBottom": 40,
          "width": 65,
        }
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#ed4848",
            "fontFamily": "nunito-extrabold",
            "fontSize": 20,
          },
          Object {
            "fontSize": 24,
          },
        ]
      }
    >
      Search for recipes
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#ed4848",
            "fontFamily": "nunito-regular",
            "fontSize": 11,
            "textAlign": "center",
          },
          Object {
            "paddingTop": 5,
            "textAlign": "left",
          },
        ]
      }
    >
      You can search by entering the ingredients you have
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#ed4848",
            "fontFamily": "nunito-regular",
            "fontSize": 11,
            "textAlign": "center",
          },
          Object {
            "paddingBottom": 15,
            "textAlign": "left",
          },
        ]
      }
    >
      or by typing in a recipe name
    </Text>
    <Image
      source={1}
      style={
        Object {
          "height": 20,
          "position": "absolute",
          "right": 50,
          "top": 217,
          "width": 20,
          "zIndex": 2,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "zIndex": 1,
          },
          Object {
            "alignSelf": "center",
            "height": 34,
            "marginBottom": 20,
            "width": 350,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderColor": "#b9b9b9",
              "borderRadius": 1,
              "borderWidth": 1,
            },
            Object {
              "borderColor": "#ed4848",
              "borderRadius": 10,
              "borderWidth": 1,
              "paddingLeft": 10,
              "paddingRight": 10,
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          containerStyle={
            Object {
              "alignSelf": "center",
              "height": 34,
              "marginBottom": 20,
              "width": 350,
            }
          }
          data={Array []}
          defaultValue=""
          flatListProps={Object {}}
          inputContainerStyle={
            Object {
              "borderColor": "#ed4848",
              "borderRadius": 10,
              "borderWidth": 1,
              "paddingLeft": 10,
              "paddingRight": 10,
            }
          }
          keyboardShouldPersistTaps="always"
          listStyle={
            Object {
              "borderWidth": 0,
            }
          }
          onChangeText={[Function]}
          onEndEditing={[Function]}
          onStartShouldSetResponderCapture={[Function]}
          rejectResponderTermination={true}
          renderItem={[Function]}
          renderSeparator={null}
          renderTextInput={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "height": 40,
                "paddingLeft": 3,
              },
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        onStartShouldSetResponderCapture={[Function]}
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#ed4848",
          "flex": 8,
          "justifyContent": "flex-start",
          "paddingBottom": 65,
          "paddingLeft": 32,
          "paddingRight": 32,
          "paddingTop": 20,
          "width": "100%",
          "zIndex": -1,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#ed4848",
            "fontFamily": "nunito-extrabold",
            "fontSize": 20,
          },
          Object {
            "color": "#fff",
            "fontSize": 24,
            "width": 350,
          },
        ]
      }
    >
      Ingredients:
    </Text>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
          "position": "absolute",
          "right": 40,
          "top": 30,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#ed4848",
              "fontFamily": "nunito-regular",
              "fontSize": 11,
              "textAlign": "center",
            },
            Object {
              "color": "#fff",
              "fontSize": 14,
            },
          ]
        }
      >
        CLEAR
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
          "paddingBottom": 5,
          "paddingTop": 15,
        }
      }
    />
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "#ed4848",
          "borderRadius": 5,
          "borderWidth": 10,
          "justifyContent": "center",
          "opacity": 1,
          "paddingBottom": 10,
          "paddingTop": 10,
          "width": 350,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "fontFamily": "nunito-bold",
              "paddingLeft": 10,
              "paddingRight": 10,
              "textAlign": "center",
            },
            Object {
              "color": "#ed4848",
            },
          ]
        }
      >
        Find Recipes
      </Text>
    </View>
  </View>
</View>
`;
